dan_layer/consensus/src/block_validations.rs

```diff
@@ -42,17 +46,50 @@ pub fn check_proposed_by_leader<TAddr: DerivableFromPublicKey, TLeaderStrategy:
     Ok(())
 }
 
-pub fn check_quorum_certificate<TAddr: NodeAddressable>(
-    _local_committee: &Committee<TAddr>,
+pub async fn check_quorum_certificate<TConsensusSpec: ConsensusSpec>(
     candidate_block: &Block,
-) -> Result<(), ProposalValidationError> {
+    vote_signing_service: &TConsensusSpec::VoteSignatureService,
+    epoch_manager: &TConsensusSpec::EpochManager,
+) -> Result<(), HotStuffError> {
+    if candidate_block.justify().epoch().as_u64() == 0 {
+        // Ignore genesis block.
+        return Ok(());
+    }
     if candidate_block.height() < candidate_block.justify().block_height() {
         return Err(ProposalValidationError::CandidateBlockNotHigherThanJustify {
             justify_block_height: candidate_block.justify().block_height(),
             candidate_block_height: candidate_block.height(),
-        });
+        }
+        .into());
+    }
+    let mut vns = vec![];
+    for signature in candidate_block.justify().signatures() {
+        let vn = epoch_manager
+            .get_validator_node_by_public_key(candidate_block.justify().epoch(), signature.public_key())
+            .await?;
+        vns.push(vn.node_hash());
     }
-    // TODO: validate QC
+    let merkle_root = epoch_manager
+        .get_validator_node_merkle_root(candidate_block.justify().epoch())
+        .await?;
+    let qc = candidate_block.justify();
+    let proof = qc.merged_proof().clone();
+    if !proof.verify_consume(&merkle_root, vns.iter().map(|hash| hash.to_vec()).collect())? {
+        return Err(ProposalValidationError::QCisNotValid { qc: qc.clone() }.into());
+    }
+
+    for (sign, leaf) in qc.signatures().iter().zip(vns.iter()) {
+        let challenge = vote_signing_service.create_challenge(leaf, qc.block_id(), &qc.decision());
+        if !sign.verify(challenge) {
+            return Err(ProposalValidationError::QCInvalidSignature { qc: qc.clone() }.into());
+        }
+    }
+    let committee_shard = epoch_manager
+        .get_committee_shard_by_validator_public_key(candidate_block.epoch(), candidate_block.proposed_by())
+        .await?;
 
+    if committee_shard.quorum_threshold() > u32::try_from(qc.signatures().len()).unwrap() {
```

Comment:
```suggestion
    if committee_shard.quorum_threshold() as usize > qc.signatures().len() {
```

Reply:
usize::from<u32> is not defined, I had to add a map_err